src/renderer/components/config/ConfigForm.tsx

```diff
@@ -0,0 +1,148 @@
+import * as React from "react";
+import * as fs from "fs";
+
+import Config, {CacheSettings, RemoteSettings, SceneSettings} from "../../Config";
+import Scene from "../../Scene";
+import APIGroup from "./APIGroup";
+import CacheGroup from "./CacheGroup";
+import EffectGroup from "../sceneDetail/EffectGroup";
+import ImageGroup from "../sceneDetail/ImageGroup";
+import TextGroup from "../sceneDetail/TextGroup";
+import TimingGroup from "../sceneDetail/TimingGroup";
+import Modal from "../ui/Modal";
+
+export default class Library extends React.Component {
+  readonly props: {
+    config: Config,
+    scenes: Array<Scene>,
+    goBack(): void,
+    updateConfig(config: Config): void,
+    onDefault(): void,
+  };
+
+  readonly state = {
+    errorMessage: "",
+    config: JSON.parse(JSON.stringify(this.props.config)), // Make a copy
+  };
+
+  render() {
+    return (
+      <div className="Config">
+        <div className="u-button-row">
+          <div className="u-abs-center">
+            <h2 className="Config__ConfigHeader">Config</h2>
+          </div>
+          <div className="u-button-row-right">
+            <div className="Config__Apply u-button u-clickable"
+                 onClick={this.applyConfig.bind(this)}>
+              Apply
+            </div>
+            <div className="Config__OK u-button u-clickable"
+                 onClick={this.onOK.bind(this)}>
+              OK
+            </div>
+          </div>
+          <div className="BackButton u-button u-clickable" onClick={this.props.goBack}>Back</div>
+          <div className="DefaultButton u-destructive u-button u-clickable"
+               onClick={this.props.onDefault.bind(this)}>
+            Reset to Defaults
+          </div>
+        </div>
+
+        <div className="Config__Content ControlGroupGroup">
+          <TimingGroup
+            scene={this.state.config.defaultScene}
+            onUpdateScene={this.onUpdateDefaultScene.bind(this)}/>
+
+          <EffectGroup
+            scene={this.state.config.defaultScene}
+            onUpdateScene={this.onUpdateDefaultScene.bind(this)}
+            allScenes={this.props.scenes}/>
+
+          <ImageGroup
+            scene={this.state.config.defaultScene}
+            onUpdateScene={this.onUpdateDefaultScene.bind(this)}/>
+
+          <TextGroup
+            scene={this.state.config.defaultScene}
+            isPlayer={false}
+            onUpdateScene={this.onUpdateDefaultScene.bind(this)}/>
+
+          <CacheGroup
+            settings={this.state.config.caching}
+            onUpdateSettings={this.onUpdateCachingSettings.bind(this)}/>
+
+          <APIGroup
+            settings={this.state.config.remoteSettings}
+            onUpdateSettings={this.onUpdateRemoteSettings.bind(this)}/>
+        </div>
+
+        {this.state.errorMessage != "" && (
+          <Modal onClose={this.onErrorClose.bind(this)} title="Error">
+            <p dangerouslySetInnerHTML={{__html: this.state.errorMessage}}/>
```

Comment:
You could also store error message as a list of strings and map them to individual `<p>` tags, instead of using the "dangerous" method.

Reply:
Derp, not sure why I didn't think of that, thanks 😋